hypershift-operator/controllers/nodepool/inplace.go

```diff
@@ -0,0 +1,402 @@
+package nodepool
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	api "github.com/openshift/hypershift/api"
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	hyperutil "github.com/openshift/hypershift/hypershift-operator/controllers/util"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	k8sutilspointer "k8s.io/utils/pointer"
+	capiv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// reconcileInPlaceUpgrade loops over all Nodes that belong to a NodePool and performs an in place upgrade if necessary.
+func (r *NodePoolReconciler) reconcileInPlaceUpgrade(ctx context.Context, hc *hyperv1.HostedCluster, nodePool *hyperv1.NodePool, machineSet *capiv1.MachineSet, targetConfigHash, targetVersion, targetConfigVersionHash string) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	// If there's no guest cluster yet return early.
+	if hc.Status.KubeConfig == nil {
+		return nil
+	}
+
+	hostedClusterClient, err := newHostedClusterClient(ctx, r.Client, hc)
+	if err != nil {
+		return fmt.Errorf("failed to create remote client: %v", err)
+	}
+
+	// Watch hosted cluster Nodes. We track the created caches, so we don't add a watcher on every reconciliation.
+	r.hostedClusterCachesTracker.Lock()
+	defer r.hostedClusterCachesTracker.Unlock()
+	if !r.hostedClusterCachesTracker.caches[client.ObjectKeyFromObject(nodePool)] {
```

Comment:
It'd be good to check how much of a memory burden is added by a separate cache per guest cluster, particularly if handling management clusters with 400+ hosted clusters,  each with several nodes.

Reply:
Added a TODO to at least index by HC instead of by NodePool. FWIW we are already doing this in the capi controller, unfortunately we need to do it here as well now to give NodePool Nodes awareness.